zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/X509AuthenticationConfig.java

```diff
@@ -343,6 +344,25 @@ private Set<String> loadCrossDomainAccessDomainNames() {
         .filter(str -> str.length() > 0).collect(Collectors.toSet());
   }
 
+  /**
+   * Check if server dedicated domain config property is set
+   * This check is only meaningful when x509 znode group acl feature is enabled
+   * @return true if a domain is set as the server's dedicated domain; false if not set
+   */
+  public boolean isZnodeGroupAclDedicatedServerEnabled() {
+    return getZnodeGroupAclServerDedicatedDomain() != null
+        && !getZnodeGroupAclServerDedicatedDomain().isEmpty();
+  }
```

Comment:
> This check is only meaningful when x509 znode group acl feature is enabled

Should we include a `isX509ZnodeGroupAclEnabled` check here?

Reply:
Can you ever enable dedicated server if znode group acl is not enabled? If the answer is no, then I think we should include the check here since this is a correctness problem before a performance problem.

Also, could you explain what you mean by duplicated checks?

Reply:
The answer is no.
In fixupACL, the code is ```X509AuthenticationConfig.getInstance().isX509ClientIdAsAclEnabled()
            && X509AuthenticationConfig.getInstance().isX509ZnodeGroupAclEnabled()
            && !X509AuthenticationConfig.getInstance().isZnodeGroupAclDedicatedServerEnabled()``` where we already check isX509ZnodeGroupAclEnabled(). So if we check once more in isZnodeGroupAclDedicatedServerEnabled(), it will be duplicated.

Reply:
Another usage of isZnodeGroupAclDedicatedServerEnabled() is in X509ZnodeGroupAclProvider, which indicates isX509ZnodeGroupAclEnabled() is true. If we include the check, it will again be a duplicated check.
I think my point is that the checks could be modularized, one method doesn't necessarily to contain all the checks because it will be difficult to reuse the methods to group into the logic that we need.

Reply:
Regarding `correctness`, I made this method name `isZnodeGroupAclDedicatedServerEnabled` which indicate that this is a sub-feature of znode group acl, so I don't think there's a problem with correctness.

Reply:
I don't think this is a performance concern, and to make our intentions clear, this is okay (this kind of stuff tends to get optimized away by the compiler anyway). If we are really concerned about these boolean checks, you can cache it in the Singleton as a boolean. (e.g. `private isZNodeGroupAclDedicatedServerEnabled` and just return that single boolean, same could apply to others, given that these are static configs).

I think it will be better to fail early if anyone tries to use dedicatedServerEnabled with any other X509 auth provider than ZNodeGroupAclProvider.

This will also help in catching potential mistakes in setting configs - if we by mistake omit the `authProvider=ZNodeGroupAclProvider` config and only set the dedicated domain, then the ZK server will be smart enough to throw an exception.

Reply:
It makes sense to fail early. I think it's ok as long as this is not a performance concern. Will update the code.